modules/viewing-the-image-pull-source.adoc

```diff
@@ -0,0 +1,44 @@
+// Module included in the following assemblies:
+//
+// *installing/validating-an-installation.adoc
+
+[id="viewing-the-image-pull-source_{context}"]
+= Viewing the image pull source
+
+You can view the source of your pulled images by using the CRI-O logs on your nodes. 
+
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+
+.Procedure
+
+* Review the CRI-O logs on a node or for the worker nodes:
++
+[source,terminal]
+----
+$  oc adm node-logs ci-ln-jb5m1vb-f76d1-cfzck-master-0 -u crio
+----
++
+.Example output
++
+The `Trying to access` log entry indicates where the images are being pulled from.
++
+[source,terminal]
+----
+...
+Aug 05 10:33:21.594996 ip-10-0-138-140.ec2.internal crio[1366]: time="2021-08-05 10:33:21.594930907Z" level=info msg="Pulling image: quay.io/openshift-release-dev/ocp-release:4.8.4-ppc64le" id=abcd713b-d0e1-4844-ac1c-474c5b60c07c name=/runtime.v1alpha2.ImageService/PullImage
+Aug 05 10:33:21.597931 ip-10-0-138-140.ec2.internal crio[1366]: time="2021-08-05 10:33:21.597839071Z" level=info msg="Trying to access \"quay.io/openshift-release-dev/ocp-release:4.8.4-ppc64le\""
```

Comment:
It might work better to provide a more complete example here, similar to https://github.com/openshift/openshift-docs/pull/35283/files#r687011124 .

Most importantly include more than one “Trying to access” line, and explain that they were tried in that order, and the last one is the one that actually succeeded (if the pull did succeed, that is).

Reply:
Ah, that’s a good point, I forgot about that: CRI-O initializes access to the same image twice, once at https://github.com/cri-o/cri-o/blob/c53e78f45ffde7a3b7a3d93d9c8c9410ca0d4fec/server/image_pull.go#L128 and once at https://github.com/cri-o/cri-o/blob/c53e78f45ffde7a3b7a3d93d9c8c9410ca0d4fec/server/image_pull.go#L229 . (FYI: @QiWang19 ). So a typical pull is indeed going to log everything twice. (If there is a _request_ for a pull, but the image already exists locally, there is one set of logs, followed by a debug-level https://github.com/cri-o/cri-o/blob/c53e78f45ffde7a3b7a3d93d9c8c9410ca0d4fec/server/image_pull.go#L157 .)

---

What I meant by “they were tried in that order, and the last one is the one that actually succeeded” is that _if_ there are multiple mirrors that are tried and fail, e.g. `quay.io/vendor/product` is configured via ICSP to be mirrored at `mirror1.local/vendor-product` and `mirror2.local/vendor-product`, but not actually available there (not copied there, or some network or permissions problem), the logs would look like this:
```
Trying to access "mirror1.local/vendor-product@digest" # Trying first mirror… fails, next attempt:
Trying to access "mirror1.local/vendor-product@digest" # Trying second mirror… fails, next attempt:
Trying to access "quay.io/vendor/product@digest" # Trying the original, succeeds
```

Or if the image were present on the second mirror, the logs would go:
```
Trying to access "mirror1.local/vendor-product@digest" # Trying first mirror… fails, next attempt:
Trying to access "mirror1.local/vendor-product@digest" # Trying second mirror… succeeds
```

Per the above note, with CRI-O, the whole log sequence would often be repeated twice, in order.

Reply:
@mtrmac Is that the exact wording in the logs? ^^

Reply:
I’d like to defer to someone who actually has a CRI-O installation around, to make sure there are no further discrepancies. Maybe @QiWang19 ?

Reply:
I will double-check the log wording.

Reply:
Checked the wording in the logs.
`docker.io/library/alpine` is configured to be mirrored at `recovery.quay.io/quay/busybox` and `recovery.quay.io/libpod/busybox`, but does not exist in the mirrors.
```
level=info msg="Pulling image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" id=8f7c20a9-c819-426c-9c3f-55ab389b5d1f name=/runtime.v1alpha2.ImageService/Pu>
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/libpod/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/libpod/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Pulled image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" 
```

`docker.io/library/alpine` is configured to be mirrored at `recovery.quay.io/quay/busybox` and `mirror.gcr.io/library/alpine`, only exists in the second mirror.

```
level=info msg="Pulling image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" id=74da5b17-db01-4e6f-a7aa-d302924d2f49 name=/runtime.v1alpha2.ImageService/Pu>
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"mirror.gcr.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"recovery.quay.io/quay/busybox@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Trying to access \"mirror.gcr.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c\""
level=info msg="Pulled image: docker.io/library/alpine@sha256:be9bdc0ef8e96dbc428dc189b31e2e3b05523d96d12ed627c37aa2936653258c" 
```

Reply:
@QiWang19 Do I understand correctly, that the log is showing that it pulled an image that is not available in the mirrored repos? Is this a bug?
Are we safe to add the note that the user should look for the _Trying to access_ statement_ in the logs? If the first example is a bug, we can add a release note if someone creates a BZ.

Reply:
@mburke5678 the first example is not a bug. If the image is not available in the mirrors, it tries the original source registry (like the quay.io/vendor/product@digest in the above comment https://github.com/openshift/openshift-docs/pull/35283#discussion_r689574193, `Trying to access "quay.io/vendor/product@digest" # Trying the original, succeeds`).  The _Trying to access_  still indicates the pull source of the image and it's safe to add the note.